### PR TITLE
salt.pillar.file_tree: add support for environments, add test

### LINKED ISSUE #44458: salt.pillar.file_tree should support environments somehow
### Description of Issue/Question

Unlike [`file_roots`](https://docs.saltstack.com/en/latest/ref/configuration/master.html#file-roots) and [`pillar_roots`](https://docs.saltstack.com/en/latest/ref/configuration/master.html#pillar-roots), the [`file_tree`](https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.file_tree.html#module-salt.pillar.file_tree) ext_pillar provides no way to specify a path per environment.  This makes developing in non-root environments on a live dev master somewhat difficult.

One option that comes to mind is to allow file_tree's `root_dir` to be relative to `pillar_roots`.  That could be done either by defining the behavior of a relative `root_dir` accordingly, or by adding a boolean argument indicating that the `root_dir` should be treated as relative.  It seems like the behavior of a relative `root_dir` is currently undefined (or at least undocumented), so I'm partial to defining the behavior of a relative `root_dir` to be relative to `pillar_roots`.

For example:
```
pillar_roots:
  base:
    - /srv/pillar
  dev:
    - /home/dev/pillar

ext_pillar:
  - file_tree:
      root_dir: relative/path
```
In this case, the base root for file_tree would be `/srv/pillar/relative/path` and the dev root for file_tree would be `/home/dev/pillar/relative/path`.  Conveniently, if someone keeps their files in `/srv/pillar_files`, they could just use a relative path of `../pillar_files`.

There's still a question of how to handle an environment with multiple directories specified in `pillar_roots`:  maybe iterate over all of them and merge the results?

If this approach seems reasonable, I'm happy to attempt a PR.  If there's a better approach, I'm not set on the above.

### Setup
From the [file_tree documentation](https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.file_tree.html#example-configuration):

```
ext_pillar:
  - file_tree:
      root_dir: /path/to/root/directory
```
### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.21.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core

### PR DESCRIPTION
**Pull request by pjcreath**
### What does this PR do?
This modifies this ext_pillar with basic support for environments by treating a relative root_dir as relative to the pillar_roots of the environment specified by pillarenv.

The behavior for absolute paths is unchanged.

Also adds a unit test for file_tree testing both old and new behavior.

### What issues does this PR fix or reference?
Resolves #44458.

### Previous Behavior
Using a relative root_dir had undocumented (and perhaps undefined) behavior. Only a single absolute root_dir could be specified for a master, preventing the use of different file_trees for different environments.

### New Behavior
A relative root_dir is treated as relative to the pillar_roots of the environment specified by pillarenv.  The behavior for an absolute root_dir is unchanged.

### Tests written?
Yes

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by pjcreath** [comment]
If you're happy with this commit, the only change that I know needs to be made is to the "versionchanged" designation in the documention. I've put "develop" as a placeholder for now.

**Comment by pjcreath** [comment]
Interesting...it looks like file_tree might already be broken on Python 3 even without this patch, but there weren't any unit tests to uncover it.  See [this Jenkins error](https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/3627/testReport/junit/unit.pillar.test_file_tree/FileTreePillarTestCase/test_absolute_path/) in code that I didn't touch:
`
  File "/testing/salt/pillar/file_tree.py", line 277, in _construct_pillar
    contents += buf
TypeError: Can't convert 'bytes' object to str implicitly
`
Should I set up a separate bug report issue + PR for that, so that it can be backported appropriately?  

Separately, [this other error](https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/3627/testReport/junit/integration.modules.test_state/StateModuleTest/test_exclude/) doesn't seem related to any of the files I touched. I'm not sure why it's failing? Maybe Jenkins is reporting the wrong test failure, since I do have 4 tests in my unit test, and presumably all 4 would fail under Python 3?

**Comment by pjcreath** [comment]
The new commit fixed the Python 3 issue in file_tree and unit test. Jenkins tests are failing outside of anything this PR touches.

File: salt/pillar/file_tree.py [hunk 1]
```diff
@@ -29,6 +29,13 @@
     Be careful when using ``follow_dir_links``, as a recursive symlink chain
     will result in unexpected results.
 
+.. versionchanged:: develop
```
Comment (terminalmage) [inline salt/pillar/file_tree.py]: Can you please change this to `Oxygen`. This is our codename for the release that the develop branch will become. Having it say `Oxygen` will make it easy for us to make the mass edits once the release number is finalized.

File: salt/pillar/file_tree.py [hunk 2]
```diff
@@ -189,6 +196,9 @@
 # Set up logging
 log = logging.getLogger(__name__)
 
+# The default option values
+__opts__ = {}
```
Comment (terminalmage) [inline salt/pillar/file_tree.py]: Why are you overwriting the packed opts dunder dictionary?

**Review by terminalmage — CHANGES_REQUESTED** [review]

File: salt/pillar/file_tree.py (hunk 1 above)
Comment (pjcreath) [reply salt/pillar/file_tree.py]: Done.

File: salt/pillar/file_tree.py (hunk 2 above)
Comment (pjcreath) [reply salt/pillar/file_tree.py]: I originally borrowed this from hg_pillar due to an error I encountered when writing the test case. Now I'm not getting the error in my test case, so it doesn't seem to be necessary any more. I've therefore removed it from the latest commit.

Note that this was only initializing the file_tree.__opts__ dictionary, not overwriting the global one.

**Comment by rallytime** [comment]
re-review from @terminalmage requested

File: salt/pillar/file_tree.py (hunk 2 above)
Comment (terminalmage) [reply salt/pillar/file_tree.py]: Yeah, I understood that. It shouldn't have been done in hg_pillar either, likely.

**Review by terminalmage — APPROVED** [review]

**Comment by rallytime** [comment]
Welcome @pjcreath! Thank you for submitting this.

**Pull request merged**